transports/quic/src/transport.rs

```
@@ -181,28 +214,59 @@ impl<P: Provider> Transport for GenTransport<P> {
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) = self.remote_multiaddr_to_socketaddr(addr.clone())?;
+        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr))?;
+
+        let endpoint_channel = match self.eligible_listener(&socket_addr) {
+            None => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            Some(listener) => listener.endpoint_channel.clone(),
+        };
+
+        let hole_puncher = hole_puncher(endpoint_channel, socket_addr, self.handshake_timeout);
+
+        let (sender, receiver) = oneshot::channel();
+
+        match self.hole_punch_attempts.entry(socket_addr) {
+            Entry::Occupied(mut sender_entry) => {
+                if sender_entry.get().is_canceled() {
+                    sender_entry.insert(sender);
+                } else {
+                    return Err(TransportError::Other(Error::HolePunchInProgress(
+                        socket_addr,
+                    )));
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(sender);
+            }
+        };
+
+        Ok(Box::pin(async move {
+            futures::pin_mut!(hole_puncher);
+            match futures::future::select(receiver, hole_puncher).await {
+                Either::Left((message, _)) => {
+                    let (inbound_peer_id, connection) = message.unwrap().await?;
+                    if inbound_peer_id != peer_id {
+                        log::warn!("inbound connection from ({socket_addr}, {inbound_peer_id} incorrectly matched with outbound hole punch for ({socket_addr}, {peer_id}).");
```

Should this be a panic instead of a warning?

---

We don't actually need to check this here, we verify that on the swarm level.

---

Can you point me to where this is checked?

---

See https://github.com/libp2p/rust-libp2p/blob/2910c985e8ab75b52d36701b9c7717ebf0fd8585/swarm/src/connection/pool.rs#L668-L685.